jax/_src/lax/lax.py

```diff
@@ -5346,3 +5346,59 @@ def convert_to(other_dtype, bint_dtype) -> bool:
 
 
 core.bint._rules = BIntRules
+
+
+def optimization_barrier(operand, /):
+  """Prevents the compiler from moving operations across the barrier.
+
+  Optimization barriers have a number of possible uses:
+
+  * An optimization barrier ensures that all inputs are evaluated before any
+    operators that depend on the barrier's outputs. This can be used to enforce
+    a particular order of operations.
+  * An optimization barrier prevents common subexpression elimination. This is
+    used by JAX to implement rematerialization.
+  * Optimization barriers prevent compiler fusions. That is, operations before
+    the barrier may not be fused into the same kernel as operations after the
+    barrier by the compiler.
+
+  JAX does not define derivative or batching rules for an optimization barrier.
```

Comment:
This seems like a problem for using it in user-level code!

In terms of value semantics, optimization_barrier is an identity function.
1. let's say that in this docstring;
2. given that we can define batching and autodiff (and all other) rules, right?

Reply:
Yeah, I erred on the side of not adding these rules for the moment, but I suspect pretty much anyone using this will have to wrap it in a custom_jvp at the moment.

So maybe I should define:
* batching: preserve the op
* jvp: don't forward it to the tangent?

Reply:
SGTM on batching.

Yeah the JVP one is interesting...

```python
# #1
primals, tangents = opt_barrier((primals, tangents))
# or, #2
primals, tangents = opt_barrier(primals), opt_barrier(tangents)
# or, #3
primals, tangents = opt_barrier(primals), tangents
```

Luckily whatever we choose as the default can be manually overridden with custom_jvp.

You think 3? I dno, maybe 1...